### ISSUE TITLE
Add support for Docker/Compose --x-networking

### ISSUE THREAD

**Issue by kenjones-cisco** [janeczku/go-dnsmasq#1]
With `--links` being deprecated and the move to using `--x-networking` the commenting out everything in `/etc/resolv.conf` means that the inter-container communication bridge is lost.

**Comment by janeczku** [comment]
@kenjones-cisco Could you explain a bit more about the implications of docker new networking and resolv.conf? Also, it should be just the nameservers that are overwritten with the other options  untouched.

**Comment by kenjones-cisco** [comment]
When using `--x-networking` with docker-compose, it will create a private bridge network and only containers on that network bridge can communicate with each other. 

An entry is placed in `/etc/hosts` with each containers: 

```
<ip address> <container name>
<ip address> <container name>.<project name>
```

And `/etc/resolv.conf` is given a namesever entry to the ip that from the same subnet as the underlying host ip.

When running go-dnsmasq as follows:

```
exec go-dnsmasq --default-resolver --hostsfile=/etc/hosts
```

I see the following:

```
consul | Dec  7 18:39:58 2ad410dc4cc2 syslog.info syslogd started: BusyBox v1.23.2
consul | ==> WARNING: Bootstrap mode enabled! Do not enable unless necessary
consul | ==> Starting raft data migration...
consul | ==> Starting Consul agent...
consul | ==> Starting Consul agent RPC...
consul | ==> Consul agent running!
consul |          Node name: '2ad410dc4cc2'
consul |         Datacenter: 'dc1'
consul |             Server: true (bootstrap: true)
consul |        Client Addr: 127.0.0.1 (HTTP: 8500, HTTPS: -1, DNS: 8600, RPC: 8400)
consul |       Cluster Addr: 172.19.0.3 (LAN: 8301, WAN: 8302)
consul |     Gossip encrypt: false, RPC-TLS: false, TLS-Incoming: false
consul |              Atlas: <disabled>
consul |
consul | ==> Log data will now stream in as it occurs:
consul |
consul |     2015/12/07 18:39:58 [INFO] raft: Node at 172.19.0.3:8300 [Follower] entering Follower state
consul |     2015/12/07 18:39:58 [INFO] serf: EventMemberJoin: 2ad410dc4cc2 172.19.0.3
consul |     2015/12/07 18:39:58 [INFO] consul: adding server 2ad410dc4cc2 (Addr: 172.19.0.3:8300) (DC: dc1)
consul |     2015/12/07 18:39:58 [INFO] serf: EventMemberJoin: 2ad410dc4cc2.dc1 172.19.0.3
consul |     2015/12/07 18:39:58 [INFO] consul: adding server 2ad410dc4cc2.dc1 (Addr: 172.19.0.3:8300) (DC: dc1)
consul |     2015/12/07 18:39:58 [ERR] agent: failed to sync remote state: No cluster leader
consul | level=info msg="Starting go-dnsmasq server 0.9.8 ..."
consul | level=info msg="Ready for queries on tcp://127.0.0.1:53 [rcache 0] Nameservers: [10.0.2.3:53]"
consul | level=info msg="Ready for queries on udp://127.0.0.1:53 [rcache 0] Nameservers: [10.0.2.3:53]"
consul |     2015/12/07 18:40:00 [WARN] raft: Heartbeat timeout reached, starting election
consul |     2015/12/07 18:40:00 [INFO] raft: Node at 172.19.0.3:8300 [Candidate] entering Candidate state
consul |     2015/12/07 18:40:00 [INFO] raft: Election won. Tally: 1
consul |     2015/12/07 18:40:00 [INFO] raft: Node at 172.19.0.3:8300 [Leader] entering Leader state
consul |     2015/12/07 18:40:00 [INFO] consul: cluster leadership acquired
consul |     2015/12/07 18:40:00 [INFO] consul: New leader elected: 2ad410dc4cc2
consul |     2015/12/07 18:40:00 [INFO] raft: Disabling EnableSingleNode (bootstrap)
consul |     2015/12/07 18:40:00 [INFO] consul: member '2ad410dc4cc2' joined, marking health alive
consul |     2015/12/07 18:40:00 [INFO] agent: Synced service 'consul'
consul | level=error msg="Failure forwarding request \"read udp 172.19.0.3:38060->10.0.2.3:53: i/o timeout\""
consul | level=error msg="Failure forwarding request \"read udp 172.19.0.3:38060->10.0.2.3:53: i/o timeout\""
consul | level=error msg="Failure forwarding request \"read udp 172.19.0.3:34722->10.0.2.3:53: i/o timeout\""
consul | level=error msg="Failure forwarding request \"read udp 172.19.0.3:34722->10.0.2.3:53: i/o timeout\""
consul | ==> Failed to check for updates: Get https://checkpoint-api.hashicorp.com/v1/check/consul?arch=amd64&os=linux&signature=aec356cc-d0eb-286e-c64f-587d42ae754b&version=0.5.2: dial tcp: lookup checkpoint-api.hashicorp.com: Try again
consul | level=error msg="Failure forwarding request \"read udp 172.19.0.3:50805->10.0.2.3:53: i/o timeout\""
consul | level=error msg="Failure forwarding request \"read udp 172.19.0.3:50805->10.0.2.3:53: i/o timeout\""
consul | level=error msg="Failure forwarding request \"read udp 172.19.0.3:42736->10.0.2.3:53: i/o timeout\""
consul | level=error msg="Failure forwarding request \"read udp 172.19.0.3:42736->10.0.2.3:53: i/o timeout\""
consul |     2015/12/07 18:40:33 [INFO] serf: EventMemberJoin: 7bb6d15c924c 172.19.0.2
consul |     2015/12/07 18:40:33 [INFO] consul: member '7bb6d15c924c' joined, marking health alive
```

When removing the option `--default-resolver` and just append `nameserver 127.0.0.1 # added by go-dnsmasq` to `/etc/resolv.conf`:

```
consul | Dec  7 18:28:02 8868ec877b4c syslog.info syslogd started: BusyBox v1.23.2
consul | ==> WARNING: Bootstrap mode enabled! Do not enable unless necessary
consul | ==> Starting raft data migration...
consul | ==> Starting Consul agent...
consul | ==> Starting Consul agent RPC...
consul | ==> Consul agent running!
consul |          Node name: '8868ec877b4c'
consul |         Datacenter: 'dc1'
consul |             Server: true (bootstrap: true)
consul |        Client Addr: 127.0.0.1 (HTTP: 8500, HTTPS: -1, DNS: 8600, RPC: 8400)
consul |       Cluster Addr: 172.19.0.3 (LAN: 8301, WAN: 8302)
consul |     Gossip encrypt: false, RPC-TLS: false, TLS-Incoming: false
consul |              Atlas: <disabled>
consul |
consul | ==> Log data will now stream in as it occurs:
consul |
consul |     2015/12/07 18:28:02 [INFO] serf: EventMemberJoin: 8868ec877b4c 172.19.0.3
consul |     2015/12/07 18:28:02 [INFO] raft: Node at 172.19.0.3:8300 [Follower] entering Follower state
consul |     2015/12/07 18:28:02 [INFO] consul: adding server 8868ec877b4c (Addr: 172.19.0.3:8300) (DC: dc1)
consul |     2015/12/07 18:28:02 [INFO] serf: EventMemberJoin: 8868ec877b4c.dc1 172.19.0.3
consul |     2015/12/07 18:28:02 [INFO] consul: adding server 8868ec877b4c.dc1 (Addr: 172.19.0.3:8300) (DC: dc1)
consul |     2015/12/07 18:28:02 [ERR] agent: failed to sync remote state: No cluster leader
consul | level=info msg="Starting go-dnsmasq server 0.9.8 ..."
consul | level=info msg="Ready for queries on tcp://127.0.0.1:53 [rcache 0] Nameservers: [10.0.2.3:53 127.0.0.1:53]"
consul | level=info msg="Ready for queries on udp://127.0.0.1:53 [rcache 0] Nameservers: [10.0.2.3:53 127.0.0.1:53]"
consul |     2015/12/07 18:28:04 [WARN] raft: Heartbeat timeout reached, starting election
consul |     2015/12/07 18:28:04 [INFO] raft: Node at 172.19.0.3:8300 [Candidate] entering Candidate state
consul |     2015/12/07 18:28:04 [INFO] raft: Election won. Tally: 1
consul |     2015/12/07 18:28:04 [INFO] raft: Node at 172.19.0.3:8300 [Leader] entering Leader state
consul |     2015/12/07 18:28:04 [INFO] consul: cluster leadership acquired
consul |     2015/12/07 18:28:04 [INFO] consul: New leader elected: 8868ec877b4c
consul |     2015/12/07 18:28:04 [INFO] raft: Disabling EnableSingleNode (bootstrap)
consul |     2015/12/07 18:28:04 [INFO] consul: member '8868ec877b4c' joined, marking health alive
consul |     2015/12/07 18:28:04 [INFO] agent: Synced service 'consul'
consul | ==> Failed to check for updates: Get https://checkpoint-api.hashicorp.com/v1/check/consul?arch=amd64&os=linux&signature=77ab184f-e3f7-8504-0eb4-420fc4b11ae1&version=0.5.2: dial tcp: lookup checkpoint-api.hashicorp.com: Try again
consul | level=error msg="Failure forwarding request \"read udp 127.0.0.1:59951->127.0.0.1:53: i/o timeout\""
consul | level=error msg="Failure forwarding request \"read udp 127.0.0.1:59951->127.0.0.1:53: i/o timeout\""
consul | level=error msg="Failure forwarding request \"read udp 127.0.0.1:51915->127.0.0.1:53: i/o timeout\""
consul | level=error msg="Failure forwarding request \"read udp 127.0.0.1:51915->127.0.0.1:53: i/o timeout\""
consul | level=error msg="Failure forwarding request \"read udp 127.0.0.1:49315->127.0.0.1:53: i/o timeout\""
consul | level=error msg="Failure forwarding request \"read udp 127.0.0.1:38433->127.0.0.1:53: i/o timeout\""
consul | level=error msg="Failure forwarding request \"read udp 127.0.0.1:33482->127.0.0.1:53: i/o timeout\""
consul | level=error msg="Failure forwarding request \"read udp 127.0.0.1:53296->127.0.0.1:53: i/o timeout\""
consul | level=error msg="Failure forwarding request \"read udp 127.0.0.1:59192->127.0.0.1:53: i/o timeout\""
consul | level=error msg="Failure forwarding request \"read udp 127.0.0.1:38497->127.0.0.1:53: i/o timeout\""
consul | level=error msg="Failure forwarding request \"read udp 127.0.0.1:51750->127.0.0.1:53: i/o timeout\""
consul | level=error msg="Failure forwarding request \"read udp 127.0.0.1:34601->127.0.0.1:53: i/o timeout\""
consul | level=error msg="Failure forwarding request \"read udp 127.0.0.1:40930->127.0.0.1:53: i/o timeout\""
consul | level=error msg="Failure forwarding request \"read udp 127.0.0.1:45618->127.0.0.1:53: i/o timeout\""
consul |     2015/12/07 18:28:37 [INFO] serf: EventMemberJoin: db021fd1d568 172.19.0.2
consul |     2015/12/07 18:28:37 [INFO] consul: member 'db021fd1d568' joined, marking health alive
consul | level=error msg="Failure forwarding request \"read udp 127.0.0.1:53729->127.0.0.1:53: i/o timeout\""
consul | level=error msg="Failure forwarding request \"read udp 127.0.0.1:38096->127.0.0.1:53: i/o timeout\""
consul | level=error msg="Failure forwarding request \"read udp 127.0.0.1:59322->127.0.0.1:53: i/o timeout\""
consul | level=error msg="Failure forwarding request \"read udp 127.0.0.1:35417->127.0.0.1:53: i/o timeout\""
consul | level=error msg="Failure forwarding request \"read udp 127.0.0.1:38237->127.0.0.1:53: i/o timeout\""
consul | level=error msg="Failure forwarding request \"read udp 127.0.0.1:58757->127.0.0.1:53: i/o timeout\""
consul | level=error msg="Failure forwarding request \"read udp 127.0.0.1:53054->127.0.0.1:53: i/o timeout\""
consul | level=error msg="Failure forwarding request \"read udp 127.0.0.1:53005->127.0.0.1:53: i/o timeout\""
consul | level=error msg="Failure forwarding request \"read udp 127.0.0.1:55966->127.0.0.1:53: i/o timeout\""
consul | level=error msg="Failure forwarding request \"read udp 127.0.0.1:34741->127.0.0.1:53: i/o timeout\""
consul | level=error msg="Failure forwarding request \"read udp 127.0.0.1:57910->127.0.0.1:53: i/o timeout\""
consul | level=error msg="Failure forwarding request \"read udp 127.0.0.1:42923->127.0.0.1:53: i/o timeout\""
consul | level=error msg="Failure forwarding request \"read udp 127.0.0.1:57127->127.0.0.1:53: i/o timeout\""
```

The errors just keep on going.

As the base image is `alpine:3.2`, I wanted to resolve the problem where I was seeing issues like:
`consul | ==> Failed to check for updates: Get https://checkpoint-api.hashicorp.com/v1/check/consul?arch=amd64&os=linux&signature=de5d9136-692b-4c94-941c-56d9c1b3399b&version=0.5.2: dial tcp: lookup checkpoint-api.hashicorp.com: Try again`

Just wanting to get around the DNS problems that exist with alpine, and still leverage the latest docker networking features that avoid having to specify which container will talk to each ahead of time and without dealing with the complexities of dnsmasq.

**Comment by janeczku** [comment]
ah i see now what you were saying , thanks

**Comment by janeczku** [comment]
> When using --x-networking with docker-compose, it will create a private bridge network and only containers on that network bridge can communicate with each other.
> And /etc/resolv.conf is given a namesever entry to the ip that from the same subnet as the underlying host ip.

I think thats the problem here: The namesever 10.0.2.3:53 is in the host's network. But the container is in it's own private bridge network. There is simply no route/connectivity from the container to the nameserver as can be observed in this line:

> Failure forwarding request \"read udp 172.19.0.3:38060->10.0.2.3:53: i/o timeout\

It looks like others are having this problem as well using consul and docker 1.9 multi-host networking: https://github.com/docker/libnetwork/issues/644

Basically you need to fix connectivity to the Consul DNS server from your containers. Go-dnsmasq can't do nothing about it. Since go-dnsmasq **is** in fact compatible with Docker 1.9 bridge networking i am closing this issue.

**Comment by kenjones-cisco** [comment]
Consul can do DNS, but I'm not using Consul for DNS, it is the KV store.

Maybe the root of the issue is the Consul does do DNS and introducing go-dnsmasq is causing things to get confused. Thanks for your feedback.

**Comment by janeczku** [comment]
@kenjones-cisco Just curious: Can you ping 10.0.2.3 from within the container?

**Comment by kenjones-cisco** [comment]
yes, the ping to 10.0.2.3 works.
